### DISCUSSION TITLE
How to run picamera2 inside Docker container

### DISCUSSION THREAD

**Discussion by hyzhak** [raspberrypi/picamera2#954] [Q&A] [answered]
Hi team, 
Any recommendations for running picamera2 in Docker Container?

Here is my Dockerfile:
```Dockerfile
FROM debian:bullseye

RUN apt update && apt install -y --no-install-recommends gnupg

RUN echo "deb http://archive.raspberrypi.org/debian/ bullseye main" > /etc/apt/sources.list.d/raspi.list \
  && apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 82B129927FA3303E

RUN apt update && apt install -y --no-install-recommends \
         python3-pip \
         python3-picamera2 \
     && apt-get clean \
     && apt-get autoremove \
     && rm -rf /var/cache/apt/archives/* \
     && rm -rf /var/lib/apt/lists/*

WORKDIR /app

COPY requirements.txt .

RUN pip install --no-cache-dir -r requirements.txt

COPY rpi_camera /app/rpi_camera

CMD ["python3", "/app/rpi_camera/main.py"]
```

docker-compose.yaml
```yaml
version: '3.8'

services:
  camera:
    build:
      context: .
      dockerfile: Dockerfile
    restart: always
    privileged: true
    devices:
      - "/dev:/dev"
```

But it fails with:
```python
INFO Camera camera_manager.cpp:297 libcamera v0.0.5+83-bde9b04f

Camera(s) not found (Do not forget to disable legacy camera with raspi-config).
Camera __init__ sequence did not complete.
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/picamera2/picamera2.py", line 242, in __init__
    self._open_camera()
  File "/usr/lib/python3/dist-packages/picamera2/picamera2.py", line 442, in _open_camera
    if not self._initialize_camera():
  File "/usr/lib/python3/dist-packages/picamera2/picamera2.py", line 406, in _initialize_camera
    raise RuntimeError("Camera(s) not found (Do not forget to disable legacy camera with raspi-config).")
RuntimeError: Camera(s) not found (Do not forget to disable legacy camera with raspi-config).
```

at the same time the same `rpi_camera/main.py` works without issues outside of Docker.

**Comment by hyzhak** [comment, marked as answer]
After some experiments, I found the optimal configuration that allows access to a camera. Here is a repository with a simplified example https://github.com/hyzhak/pi-camera-in-docker

**Reply by DominiquePaul** [reply]
Thank you for sharing this. On which raspberry pi and OS did you run this?

**Comment by rexkoh425** [comment]
This might be the only missing line in your docker-compose.yml for anyone referencing this in the future

volumes:
      - /run/udev:/run/udev:ro

**Comment by HudsonReynolds2** [comment]
As of Jan 30 2026, I got this working on a Pi 5 with a Pi Camera Module V3. I was in development mode.

Dockerfile:
FROM balenalib/raspberrypi5-debian:bookworm

# Enable udev
ENV UDEV=1

RUN apt update && apt install -y --no-install-recommends gnupg
RUN echo "deb http://archive.raspberrypi.org/debian/ bookworm main" > /etc/apt/sources.list.d/raspi.list \
  && apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 82B129927FA3303E
RUN apt update && apt install -y --no-install-recommends \
         python3-pip \
         python3-picamera2 \
     && apt-get clean \
     && apt-get autoremove \
     && rm -rf /var/cache/apt/archives/* \
     && rm -rf /var/lib/apt/lists/*

WORKDIR /app
COPY requirements.txt .
RUN pip install --break-system-packages --no-cache-dir -r requirements.txt
COPY pi_camera_in_docker /app/pi_camera_in_docker
CMD ["python3", "/app/pi_camera_in_docker/main.py"]


docker-compose.yml:
version: '2.1'
services:
  camera:
    build:
      context: .
    devices:
      - "/dev:/dev"
    labels:
      io.balena.features.sysfs: '1'
    ports:
      - "8000:8000"
    restart: always
    privileged: true
    network_mode: host
    
 
<img width="557" height="119" alt="image" src="https://github.com/user-attachments/assets/48b45464-764e-40cc-84d7-0f05f59c16f6" />
